## Notebook com todos os desafios propostos

### Desafio 1

1) Observe o trecho de código abaixo: int INDICE = 13, SOMA = 0, K = 0;

Enquanto K < INDICE faça { K = K + 1; SOMA = SOMA + K; }

Imprimir(SOMA);

Ao final do processamento, qual será o valor da variável SOMA?

In [1]:
def calcular_soma():
    INDICE = 13
    SOMA = 0
    K = 0

    while K < INDICE:
        K += 1
        SOMA += K

    print(f"O valor da variável SOMA é: {SOMA}")

calcular_soma()

O valor da variável SOMA é: 91


### RESPOSTA: O valor da variável SOMA será 91


### Desafio 2

2) Dado a sequência de Fibonacci, onde se inicia por 0 e 1 e o próximo valor sempre será a soma dos 2 valores anteriores (exemplo: 0, 1, 1, 2, 3, 5, 8, 13, 21, 34...), escreva um programa na linguagem que desejar onde, informado um número, ele calcule a sequência de Fibonacci e retorne uma mensagem avisando se o número informado pertence ou não a sequência.

IMPORTANTE: Esse número pode ser informado através de qualquer entrada de sua preferência ou pode ser previamente definido no código;

In [2]:
def verificar_fibonacci(numero):
    a, b = 0, 1

    while b <= numero:
        if b == numero or a == numero:
            print(f"O número {numero} pertence à sequência de Fibonacci.")
            return
        a, b = b, a + b

    print(f"O número {numero} não pertence à sequência de Fibonacci.")

verificar_fibonacci(5)
verificar_fibonacci(21)
verificar_fibonacci(33)
verificar_fibonacci(35)
verificar_fibonacci(100)

O número 5 pertence à sequência de Fibonacci.
O número 21 pertence à sequência de Fibonacci.
O número 33 não pertence à sequência de Fibonacci.
O número 35 não pertence à sequência de Fibonacci.
O número 100 não pertence à sequência de Fibonacci.


### Resposta (Com os números propostos, para mudar essa resposta basta substituir algum número pelo número desejado): 

O número 5 pertence à sequência de Fibonacci.

O número 21 pertence à sequência de Fibonacci.

O número 33 não pertence à sequência de Fibonacci.

O número 35 não pertence à sequência de Fibonacci.

O número 100 não pertence à sequência de Fibonacci.

### Desafio 3

3) Dado um vetor que guarda o valor de faturamento diário de uma distribuidora, faça um programa, na linguagem que desejar, que calcule e retorne:
• O menor valor de faturamento ocorrido em um dia do mês;
• O maior valor de faturamento ocorrido em um dia do mês;
• Número de dias no mês em que o valor de faturamento diário foi superior à média mensal.

IMPORTANTE:
a) Usar o json ou xml disponível como fonte dos dados do faturamento mensal;
b) Podem existir dias sem faturamento, como nos finais de semana e feriados. Estes dias devem ser ignorados no cálculo da média;

In [1]:
def analisar_faturamento(faturamento_diario):
    dias_com_faturamento = [faturamento['valor'] for faturamento in faturamento_diario if faturamento['valor'] > 0]

    menor_valor = min(dias_com_faturamento)
    maior_valor = max(dias_com_faturamento)

    media = sum(dias_com_faturamento) / len(dias_com_faturamento)

    dias_acima_da_media = len([d for d in dias_com_faturamento if d > media])

    print(f"Menor valor de faturamento: R${menor_valor:.2f}")
    print(f"Maior valor de faturamento: R${maior_valor:.2f}")
    print(f"Número de dias com faturamento acima da média: {dias_acima_da_media}")

faturamento_diario = [
    {"dia": 1, "valor": 22174.1664},
    {"dia": 2, "valor": 24537.6698},
    {"dia": 3, "valor": 26139.6134},
    {"dia": 4, "valor": 0.0},
    {"dia": 5, "valor": 0.0},
    {"dia": 6, "valor": 26742.6612},
    {"dia": 7, "valor": 0.0},
    {"dia": 8, "valor": 42889.2258},
    {"dia": 9, "valor": 46251.174},
    {"dia": 10, "valor": 11191.4722},
    {"dia": 11, "valor": 0.0},
    {"dia": 12, "valor": 0.0},
    {"dia": 13, "valor": 3847.4823},
    {"dia": 14, "valor": 373.7838},
    {"dia": 15, "valor": 2659.7563},
    {"dia": 16, "valor": 48924.2448},
    {"dia": 17, "valor": 18419.2614},
    {"dia": 18, "valor": 0.0},
    {"dia": 19, "valor": 0.0},
    {"dia": 20, "valor": 35240.1826},
    {"dia": 21, "valor": 43829.1667},
    {"dia": 22, "valor": 18235.6852},
    {"dia": 23, "valor": 4355.0662},
    {"dia": 24, "valor": 13327.1025},
    {"dia": 25, "valor": 0.0},
    {"dia": 26, "valor": 0.0},
    {"dia": 27, "valor": 25681.8318},
    {"dia": 28, "valor": 1718.1221},
    {"dia": 29, "valor": 13220.495},
    {"dia": 30, "valor": 8414.61}
]

analisar_faturamento(faturamento_diario)

Menor valor de faturamento: R$373.78
Maior valor de faturamento: R$48924.24
Número de dias com faturamento acima da média: 10


# RESPOSTA (Com o faturamento proposto): 

Menor valor de faturamento: R$373.78

Maior valor de faturamento: R$48924.24

Número de dias com faturamento acima da média: 10

### Desafio 4

4) Dado o valor de faturamento mensal de uma distribuidora, detalhado por estado:
• SP – R$67.836,43
• RJ – R$36.678,66
• MG – R$29.229,88
• ES – R$27.165,48
• Outros – R$19.849,53

Escreva um programa na linguagem que desejar onde calcule o percentual de representação que cada estado teve dentro do valor total mensal da distribuidora.  


In [4]:
def calcular_percentual_por_estado():
    faturamento_por_estado = {
        "SP": 67836.43,
        "RJ": 36678.66,
        "MG": 29229.88,
        "ES": 27165.48,
        "Outros": 19849.53,
    }

    total = sum(faturamento_por_estado.values())

    for estado, valor in faturamento_por_estado.items():
        percentual = (valor / total) * 100
        print(f"{estado}: {percentual:.2f}%")

calcular_percentual_por_estado()

SP: 37.53%
RJ: 20.29%
MG: 16.17%
ES: 15.03%
Outros: 10.98%


### Resposta: 

SP: 37.53%

RJ: 20.29%

MG: 16.17%

ES: 15.03%

Outros: 10.98%

### Desafio 5: 

5) Escreva um programa que inverta os caracteres de um string.

IMPORTANTE:

a) Essa string pode ser informada através de qualquer entrada de sua preferência ou pode ser previamente definida no código;

b) Evite usar funções prontas, como, por exemplo, reverse;

In [6]:
def inverter_string(texto):
    invertida = ""

    for i in range(len(texto) - 1, -1, -1):
        invertida += texto[i]

    print(f"Texto original: {texto}")
    print(f"Texto invertido: {invertida}")

inverter_string("Fortaleza é uma cidade linda")


Texto original: Fortaleza é uma cidade linda
Texto invertido: adnil edadic amu é azelatroF
